### ISSUE TITLE
Force Universal to wait for component promises to resolve before rendering.

### ISSUE THREAD

**Issue by claridgicus** [angular/angular#51016]
# 🚀 Feature request

#### What modules are relevant for this feature request?

- [ ] builders
- [X] common
- [X] express-engine

### Description

I have an angular universal instance that requires data from a CMS (prismic) to render the majority of the page, (theme data, menu systems, etc). I use the CMS's SDK to make all requests, as you might expect when working with a third-party software vendor. So this request does not use the angular HTTP client. In some instances, what I've noticed is that sporadically universal will have either waited for this request or skipped it entirely in the rendered HTML.

Is there a way to mark specific Promises/HTTP requests as required for the page to bootstrap correctly? As I understand wrapping a third-party libraries request in a macrotask is no longer an option. 

### Describe the solution you'd like

Documentation around this topic / Some method of wrapping such an essential promise from a third-party library that will force Universal to wait for the result of the HTTP request before rendering its HTML.

### Ideal Solution

Potentially some lifecycle hook or other service that allows me to say, "When you've hit this point you're allowed to respond to the request with what you have"

### What have I tried so far

I've tried, moving it between the service and the bootstrapping component, loading with observables, promises and anything that comes to mind.

Half a solution to my problem would be chucking the theme object into some KV and passing it in through the provider's section of the bootstrapping process, but that does not help requests that are requesting things like "page content"

The only other solution I can think of would be to implement some KV cache that any request by this library (prismicio/client) would have to check before completing, lowering the frequency of this issue occurring but not resolving it.

I have also tried implementing and tinkering with this snippet here. however, my issue is still occurring.  https://gist.github.com/sparebytes/e2bc438e3cfca7f6687f1d61287f8d72

**Comment by alan-agius4** [comment]
Transferring to the FW as this is more of a framework issue.

Can you please provide the output of `ng version`?

**Comment by alan-agius4** [comment]
And to confirm are you using zone.js 0.13.1?

**Comment by claridgicus** [comment]
@alan-agius4

That would be correct :)

**Comment by claridgicus** [comment]
@AndrewKushnir I have clarified - for the sake of pushing my ticket along.

**Comment by claridgicus** [comment]
This seems to be specifically when not using the httpClient to make HTTP requests, which would be how the majority of out-of-the-box libraries work.

I made a minimal repro instance of angular universal and found it to be a pretty common occurrence with any 3rd party library I was using.

I've managed to "solve" this problem by including this particular service
[Here](https://stackoverflow.com/questions/47151811/server-does-not-wait-till-http-call-completes-before-rendering-angular-4-serve)
With some small amendments to make it compatible with 16.

However, in the interest of DX I think this is a pretty easy thing for the framework to handle - There should also be some level of documentation around this aspect of universal.
